packages/api/src/graphql/resolvers/CandleResolver.ts

```diff
@@ -72,6 +79,93 @@ const groupPricesByInterval = (
   return candles;
 };
 
+const getTrailingAveragePricesByInterval = (
+  prices: ResourcePricePoint[],
+  trailingIntervalSeconds: number,
+  intervalSeconds: number,
+  startTimestamp: number,
+  endTimestamp: number,
+  lastKnownPrice?: string
+): CandleType[] => {
+  const candles: CandleType[] = [];
+
+  // If we have no prices and no reference price, return empty array
+  if (prices.length === 0 && !lastKnownPrice) return [];
+
+  // Normalize timestamps to interval boundaries
+  const normalizedStartTimestamp =
+    Math.floor(startTimestamp / intervalSeconds) * intervalSeconds;
+  const normalizedEndTimestamp =
+    Math.floor(endTimestamp / intervalSeconds) * intervalSeconds;
+
+  // Initialize lastClose with lastKnownPrice if available, otherwise use first price
+  let lastClose = lastKnownPrice || prices[0].value;
+
+  const orderedPrices = prices.sort((a, b) => a.timestamp - b.timestamp);
+  let lastStartIdx = 0;
+  let lastEndIdx = 0;
+  let startIdx = 0;
+  let endIdx = 0;
+
+  let totalGasUsed: bigint = 0n;
+  let totalBaseFeesPaid: bigint = 0n;
+
+  for (
+    let timestamp = normalizedStartTimestamp;
+    timestamp <= normalizedEndTimestamp;
+    timestamp += intervalSeconds
+  ) {
+    // get the indexes for the start and end of the interval
+    startIdx = orderedPrices.findIndex(
+      (p) => p.timestamp >= timestamp - trailingIntervalSeconds
+    );
+    endIdx = orderedPrices.findIndex((p) => p.timestamp > timestamp); // notice is the next item, we need to correct it later
+
+    // Remove from the sliding window trailing average the prices that are no longer in the interval
+    if (startIdx != -1) {
+      for (let i = lastStartIdx; i <= startIdx; i++) {
+        totalGasUsed -= BigInt(orderedPrices[i].used);
+        totalBaseFeesPaid -= BigInt(orderedPrices[i].feePaid);
+      }
+    }
+    lastStartIdx = startIdx;
+
+    // if found and not previous endIdx, correct the +1 offset of the endIdx (since we found the next item)
+    if (endIdx != -1 && endIdx > lastEndIdx) {
+      endIdx--;
+    }
+
+    // If not found, use the last index of the orderedPrices array
+    if (endIdx == -1) {
+      endIdx = orderedPrices.length - 1;
+    }
+
+    // Add to the sliding window trailing average the prices that are now in the interval
+    for (let i = lastEndIdx; i <= endIdx; i++) {
```

Comment:
we could try to make the calculations faster; right now the runtime is approx. O(n^2) since we have a nested loop. If the performance is not too big of a concern we should be good though